### PR TITLE
Install StyleCop.Analyzers

### PR DESCRIPTION
**Pull request by NextTurn**
You probably won't like many rules enabled by default.

### CODE REVIEW

**Comment by NextTurn** [comment]
Some rules aren't working with new C# language features. They need to be fixed in StyleCop repo first.

File: src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdHeap.cs [hunk 1]
```diff
@@ -173,11 +171,12 @@ private IReadOnlyList<ClrSegment> Initialize()
         {
             lock (_sync)
             {
-                _helpers.CreateSegments(this,
-                                        out IReadOnlyList<ClrSegment> segments,
-                                        out IReadOnlyList<AllocationContext> allocContext,
-                                        out IReadOnlyList<FinalizerQueueSegment> fqRoots,
-                                        out IReadOnlyList<FinalizerQueueSegment> fqObjects);
+                _helpers.CreateSegments(
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdHeap.cs]: I dislike this refactor, is there a way to turn this one off?

**Comment by leculver** [comment]
The mechanical changes overall look good so far, minus one question.

File: src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdHeap.cs (hunk 1 above)
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdHeap.cs]: Aligning by opening parentheses will result in inconsistencies...

File: src/Microsoft.Diagnostics.Runtime.Utilities/Microsoft.Diagnostics.Runtime.Utilities.csproj [hunk 2]
```diff
@@ -1,25 +1,11 @@
...
-    <Description>.</Description>
-    <PackageReleaseNotes></PackageReleaseNotes>
-    <PackageTags></PackageTags>
-    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <CheckForOverflowUnderflow>false</CheckForOverflowUnderflow>
-    <AssemblyOriginatorKeyFile>$(MSBuildThisFileDirectory)\..\..\eng\Microsoft.Diagnostics.Runtime.snk</AssemblyOriginatorKeyFile>
-    <RepositoryUrl>https://github.com/Microsoft/clrmd</RepositoryUrl>
-  </PropertyGroup>
-  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|net471|AnyCPU'">
-    <DefineConstants>TRACE;DEBUG;NET45</DefineConstants>
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime.Utilities/Microsoft.Diagnostics.Runtime.Utilities.csproj]: Does this remove the DEBUG constant or is that implied somehow?

File: src/Microsoft.Diagnostics.Runtime.Tests/Microsoft.Diagnostics.Runtime.Tests.csproj [hunk 3]
```diff
@@ -1,8 +1,7 @@
-﻿<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <TargetFrameworks>netcoreapp3.0</TargetFrameworks>
-    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <IsPackable>false</IsPackable>
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime.Tests/Microsoft.Diagnostics.Runtime.Tests.csproj]: Why is this set to false?

I don't have any objection if there's reason, just wanted to understand here.

File: Directory.Build.props [hunk 4]
```diff
@@ -7,13 +7,10 @@
   <PropertyGroup>
     <LangVersion>Latest</LangVersion>
     <Features>strict</Features>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
   </PropertyGroup>
 
-  <PropertyGroup Condition="'$(Configuration)' == 'Debug'">
-    <CheckForOverflowUnderflow>true</CheckForOverflowUnderflow>
```
Comment (leculver) [inline Directory.Build.props]: This was intentional.  We do not want overflow/underflow to interfere with the library but I did want actual overflow issues to surface in test runs or when debugging the library.  Was this changed on purpose?

File: src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdHeap.cs (hunk 1 above)
Comment (leculver) [reply src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdHeap.cs]: I suppose that's fine, but it's going to be a hard habit to break after 20 years of doing it the other way.  =)

File: src/Microsoft.Diagnostics.Runtime.Utilities/Microsoft.Diagnostics.Runtime.Utilities.csproj (hunk 2 above)
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime.Utilities/Microsoft.Diagnostics.Runtime.Utilities.csproj]: It's implied by the `Debug` configuration.

File: src/Microsoft.Diagnostics.Runtime.Tests/Microsoft.Diagnostics.Runtime.Tests.csproj (hunk 3 above)
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime.Tests/Microsoft.Diagnostics.Runtime.Tests.csproj]: I saw this is set to false for UnitTests.csproj so I synced them.

File: Directory.Build.props (hunk 4 above)
Comment (NextTurn) [reply Directory.Build.props]: I'll revert this.

**Comment by NextTurn** [comment]
Functionality changes are unintentional. If there is anything beyond formatting (like the overflow/underflow one) please do leave a comment.

File: Directory.Build.props (hunk 4 above)
Comment (NextTurn) [reply Directory.Build.props]: The current configurations override this property to always false in non-test libraries.

Comment (NextTurn) [reply Directory.Build.props]: Enabling overflow checks broke a bunch of tests 😅. That's a real functionality change which should be saved for later.

File: src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdHeap.cs (hunk 1 above)
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdHeap.cs]: Also, is it a habit to spell out .NET as .Net? It should be all capitalized.

[.NET Glossary](https://github.com/dotnet/docs/blob/master/docs/standard/glossary.md#net)

[Microsoft Style Guide](https://github.com/MicrosoftDocs/microsoft-style-guide/blob/master/styleguide/a-z-word-list-term-collections/n/net.md)

File: Directory.Build.props (hunk 4 above)
Comment (leculver) [reply Directory.Build.props]: haha that's pretty interesting...  I should look into that at some point.

**Pull request merged**